File: src/__tests__/protocol-fixtures.test.ts
import fs from "node:fs";
import path from "node:path";
import { describe, expect, it } from "vitest";
import type { z } from "zod";
import {
	ConsoleCloseSessionResultSchema,
	ConsoleCreateSessionResultSchema,
	ConsoleForgetResultSchema,
	ConsoleListDirsResultSchema,
	ConsoleOpSchema,
	ConsolePeekResultSchema,
	ConsoleSendResultSchema,
	InboxRowSchema,
	OwnerOpSchema,
	PlanesReadResultSchema,
	PlanesReadValueSchema,
} from "../shared/schemas.js";
import { MailboxEntrySchema } from "../shared/schemasConsoleOp.js";

const FIXTURES = path.join(__dirname, "../../tests/fixtures/protocol");

const SCHEMAS: Record<string, z.ZodType> = {
	ConsoleOp: ConsoleOpSchema,
	MailboxEntry: MailboxEntrySchema,
	OwnerOp: OwnerOpSchema,
	InboxRow: InboxRowSchema,
	PlanesReadValue: PlanesReadValueSchema,
	PlanesReadResult: PlanesReadResultSchema,
	ConsoleSendResult: ConsoleSendResultSchema,
	ConsoleCreateSessionResult: ConsoleCreateSessionResultSchema,
	ConsoleCloseSessionResult: ConsoleCloseSessionResultSchema,
	ConsoleForgetResult: ConsoleForgetResultSchema,
	ConsoleListDirsResult: ConsoleListDirsResultSchema,
	ConsolePeekResult: ConsolePeekResultSchema,
};

interface ManifestEntry {
	file: string;
	schema: string;
	expect: "pass" | "fail";
}

function fixture(name: string): unknown {
	return JSON.parse(fs.readFileSync(path.join(FIXTURES, name), "utf8"));
}

const manifest = (fixture("_manifest.json") as { fixtures: ManifestEntry[] }).fixtures;

describe("protocol fixtures", () => {
	it("manifest covers exactly the fixture files on disk", () => {
		const onDisk = fs
			.readdirSync(FIXTURES)
			.filter((name) => name.endsWith(".json") && name !== "_manifest.json")
			.sort();
		expect(manifest.map((entry) => entry.file).sort()).toEqual(onDisk);
	});

	it("manifest schemas all resolve", () => {
		for (const entry of manifest) {
			expect(SCHEMAS[entry.schema], `unknown schema ${entry.schema} for ${entry.file}`).toBeDefined();
		}
	});

	it.each(manifest.map((entry) => [entry.file, entry] as const))("%s matches its manifest entry", (_, entry) => {
		const result = SCHEMAS[entry.schema].safeParse(fixture(entry.file));
		expect(result.success).toBe(entry.expect === "pass");
	});

	it("preserves large integer fields", () => {
		expect(OwnerOpSchema.parse(fixture("owner-op.json")).at).toBeGreaterThan(2 ** 31);
		expect(PlanesReadValueSchema.parse(fixture("planes-read-value.json")).known.presence.version).toBe(4294967296);
	});

	it("accepts additive fields without exposing them", () => {
		const op = { ...(fixture("console-op-create-session.json") as object), futureField: true };
		const result = ConsoleOpSchema.parse(op);
		expect(result).not.toHaveProperty("futureField");
	});

	it("keeps the retained result shapes covered", () => {
		expect(ConsoleSendResultSchema.parse(fixture("send-result.json")).status).toBe("delivered");
		expect(ConsoleCreateSessionResultSchema.parse(fixture("create-session-result-pending.json")).status).toBe(
			"pending",
		);
		expect(ConsoleCloseSessionResultSchema.parse(fixture("close-session-result.json"))).toBeDefined();
		expect(ConsoleForgetResultSchema.parse(fixture("forget-result-release.json")).boardDisposition).toBe("release");
		expect(ConsoleListDirsResultSchema.parse(fixture("list-dirs-result.json"))).toBeDefined();
		expect(InboxRowSchema.parse(fixture("inbox-row.json")).seq).toBe(1);
	});
});

File: android/app/src/test/java/com/atelier_nyaarium/switchboard/ProtocolFixturesTest.kt
package com.atelier_nyaarium.switchboard

import com.atelier_nyaarium.switchboard.proto.ConsoleCloseSessionResult
import com.atelier_nyaarium.switchboard.proto.ConsoleCreateSessionResult
import com.atelier_nyaarium.switchboard.proto.ConsoleForgetResult
import com.atelier_nyaarium.switchboard.proto.ConsoleListDirsResult
import com.atelier_nyaarium.switchboard.proto.ConsoleOp
import com.atelier_nyaarium.switchboard.proto.ConsolePeekResult
import com.atelier_nyaarium.switchboard.proto.ConsoleSendResult
import com.atelier_nyaarium.switchboard.proto.InboxRow
import com.atelier_nyaarium.switchboard.proto.MailboxEntry
import com.atelier_nyaarium.switchboard.proto.OwnerOp
import com.atelier_nyaarium.switchboard.proto.PlanesReadResult
import com.atelier_nyaarium.switchboard.proto.PlanesReadValue
import com.atelier_nyaarium.switchboard.proto.WorkspaceFileMutationAnswer
import com.atelier_nyaarium.switchboard.proto.WorkspaceSaveSpanAnswer
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ProtocolFixturesTest {
	private fun fixture(name: String): String =
		javaClass.classLoader?.getResourceAsStream("protocol/$name")?.bufferedReader()?.use { it.readText() }
			?: error("missing protocol fixture: $name")

	private fun decodeAs(schema: String, body: String): Any =
		when (schema) {
			"ConsoleOp" -> wireJson.decodeFromString<ConsoleOp>(body)
			"OwnerOp" -> wireJson.decodeFromString<OwnerOp>(body)
			"InboxRow" -> wireJson.decodeFromString<InboxRow>(body)
			"MailboxEntry" -> wireJson.decodeFromString<MailboxEntry>(body)
			"PlanesReadValue" -> wireJson.decodeFromString<PlanesReadValue>(body)
			"PlanesReadResult" -> wireJson.decodeFromString<PlanesReadResult>(body)
			"ConsoleSendResult" -> wireJson.decodeFromString<ConsoleSendResult>(body)
			"ConsoleCreateSessionResult" -> wireJson.decodeFromString<ConsoleCreateSessionResult>(body)
			"ConsoleCloseSessionResult" -> wireJson.decodeFromString<ConsoleCloseSessionResult>(body)
			"ConsoleForgetResult" -> wireJson.decodeFromString<ConsoleForgetResult>(body)
			"ConsoleListDirsResult" -> wireJson.decodeFromString<ConsoleListDirsResult>(body)
			"ConsolePeekResult" -> wireJson.decodeFromString<ConsolePeekResult>(body)
			"WorkspaceSaveSpanAnswer" -> wireJson.decodeFromString<WorkspaceSaveSpanAnswer>(body)
			"WorkspaceFileMutationAnswer" -> wireJson.decodeFromString<WorkspaceFileMutationAnswer>(body)
			else -> error("unknown manifest schema: $schema")
		}

	@Test
	fun everyManifestFixtureExistsAndDecodesAsDeclared() {
		val manifest = wireJson.parseToJsonElement(fixture("_manifest.json")).jsonObject["fixtures"]!!.jsonArray
		assertTrue("manifest must not be empty", manifest.isNotEmpty())
		for (entry in manifest) {
			val obj = entry.jsonObject
			val file = obj["file"]!!.jsonPrimitive.content
			val schema = obj["schema"]!!.jsonPrimitive.content
			val expectPass = obj["expect"]!!.jsonPrimitive.content == "pass"
			val decoded = runCatching { decodeAs(schema, fixture(file)) }
			assertEquals("$file decode result", expectPass, decoded.isSuccess)
			if (expectPass) assertNotNull("$file decoded as $schema", decoded.getOrNull())
		}
	}

	@Test
	fun preservesLargeIntegers() {
		val owner = wireJson.decodeFromString<OwnerOp>(fixture("owner-op.json"))
		assertTrue(owner.at > Int.MAX_VALUE.toLong())
		val planes = wireJson.decodeFromString<PlanesReadValue>(fixture("planes-read-value.json"))
		assertEquals(4294967296L, planes.known["presence"]!!.jsonObject["version"]!!.jsonPrimitive.long)
	}

	@Test
	fun toleratesAdditiveUnknownFields() {
		val original = wireJson.parseToJsonElement(fixture("console-op-send.json")).jsonObject
		val withUnknown = JsonObject(original + ("futureField" to JsonPrimitive(true)))
		val decoded = wireJson.decodeFromJsonElement<ConsoleOp>(withUnknown)
		assertTrue(decoded is ConsoleOp.Send)
		assertEquals("status?", (decoded as ConsoleOp.Send).body)
	}

	@Test
	fun retainsResultShapes() {
		assertEquals("delivered", wireJson.decodeFromString<ConsoleSendResult>(fixture("send-result.json")).status)
		assertEquals(
			"pending",
			wireJson.decodeFromString<ConsoleCreateSessionResult>(fixture("create-session-result-pending.json")).status,
		)
		assertTrue(wireJson.decodeFromString<ConsoleCloseSessionResult>(fixture("close-session-result.json")).closed)
		assertEquals(
			"release",
			wireJson.decodeFromString<ConsoleForgetResult>(fixture("forget-result-release.json")).boardDisposition,
		)
		assertNotNull(wireJson.decodeFromString<ConsoleListDirsResult>(fixture("list-dirs-result.json")).entries)
		assertTrue(wireJson.decodeFromString<ConsolePeekResult>(fixture("peek-result-unchanged.json")).unchanged == true)
		assertEquals(1L, wireJson.decodeFromString<InboxRow>(fixture("inbox-row.json")).seq)
	}
}
